llvm/lib/Transforms/Vectorize/LoopVectorize.cpp

```diff
@@ -5544,14 +5544,28 @@ InstructionCost LoopVectorizationCostModel::computePredInstDiscount(
 InstructionCost LoopVectorizationCostModel::expectedCost(ElementCount VF) {
   InstructionCost Cost;
 
+  // If with the given VF loop gets fully unrolled, ignore the costs of
+  // comparison and induction instructions, as they'll get simplified away
+  SmallPtrSet<const Value *, 16> ValuesToIgnoreForVF;
+  auto TC = PSE.getSE()->getSmallConstantTripCount(TheLoop);
+  auto *Cmp = TheLoop->getLatchCmpInst();
+  if (Cmp && TC == VF.getKnownMinValue()) {
+    ValuesToIgnoreForVF.insert(Cmp);
+    for (const auto &[IV, IndDesc] : Legal->getInductionVars()) {
```

Comment:
I don't think this looks right, since there could be outside users of the induction variables and the increment will remain even after the loop is removed. I think you need to make sure that the only use of the increment is the phi in the loop header.

Reply:
Yes indeed. I added the check that the next iteration value is used only by induction PHI or latch comparison.